platform/podjs/src/browserPod.ts

```diff
@@ -11,67 +11,179 @@ import type {
+
 const NAV_FRAME_ID = "polyNavFrame";
 const NAV_DEFAULT_BACKGROUND_COLOR = "#ffffff";
 const NAV_DARK_FOREGROUND_COLOR = "#000000";
 const NAV_LIGHT_FOREGROUND_COLOR = "#ffffff";
 
 const MANIFEST_DATA = window.manifestData;
 
-class LocalStoragePolyIn implements PolyIn {
-    private static readonly storageKey = "polyInStore";
-    private store = JSON.parse(
-        localStorage.getItem(LocalStoragePolyIn.storageKey) || "[]"
-    );
+// Previously, data were stored in localStorage. IndexedDB is more flexible and
+// efficient, and most importantly in our case can store larger amount of data,
+// while the 2MB limit of localStorage is no longer sufficient. This function
+// migrates any data in localStorage to not disrupt the developer expirience.
+// TODO: Please remove the migration code after Oct 2022
+async function migrateData(db: IDBDatabase): Promise<IDBDatabase> {
+    const polyInData = localStorage.getItem("polyInStore");
+    const polyOutData = localStorage.getItem("files");
+
+    if (polyInData || polyOutData) {
+        const files: FileInfo[] = [];
+        const quads = polyInData
+            ? JSON.parse(polyInData).map(RDFString.quadToStringQuad)
+            : [];
+
+        if (polyOutData) {
+            for (const [, file] of JSON.parse(polyOutData)) {
+                const dataUrl = localStorage.getItem(file.id);
+                if (dataUrl) {
+                    files.push({
+                        id: file.id,
+                        name: file.name,
+                        time: new Date(file.time),
+                        blob: await (await fetch(dataUrl)).blob(),
+                    });
+                }
+            }
+        }
+
+        await new Promise((resolve, reject) => {
+            const storeNames = [OBJECT_STORE_POLY_IN, OBJECT_STORE_POLY_OUT];
+            const tx = db.transaction(storeNames, "readwrite");
+            const polyInStore = tx.objectStore(OBJECT_STORE_POLY_IN);
+            for (const quad of quads) polyInStore.add(quad);
+            const polyOutStore = tx.objectStore(OBJECT_STORE_POLY_OUT);
+            for (const file of files) polyOutStore.add(file);
+            tx.oncomplete = () => resolve(null);
+            tx.onerror = tx.onabort = () => reject(tx.error);
+        });
+
+        localStorage.clear();
+    }
+
+    return db;
+}
+
+function openDatabase(): Promise<IDBDatabase> {
+    return new Promise((resolve, reject) => {
+        const request = indexedDB.open(DB_NAME, DB_VERSION);
+
+        request.onupgradeneeded = () => {
+            const db = request.result;
+            const polyInStore = db.createObjectStore(OBJECT_STORE_POLY_IN, {
+                keyPath: QUAD_KEYS,
+            });
+            for (let mask = 1; mask < (1 << QUAD_KEYS.length) - 1; mask++) {
+                const keyPath = QUAD_KEYS.filter((_, idx) => (mask >> idx) & 1);
+                polyInStore.createIndex(keyPath.join("-"), keyPath);
+            }
```

Comment:
This part took me a while to parse, I think I'd have found it easier to understand what it's for, e.g. by having it in a separate function. Something like:

```
createIndexesForKeys(polyInStore, QUAD_KEYS)
```

Reply:
As we already discussed this on Slack, let me just sum up our discussion for reference:

Since `PolyIn.match` queries `Quad`s by any subset of `subject`, `predicate`, `object`,  we have to create individual indexes using a composite key for each of those combinations.

If you think about it, combinations can be easily represented as binary numbers:

| `subject` | `predicate` | `object` | binary representation | key path                             |
|-----------|-------------|----------|-----------------------|--------------------------------------|
| yes (1)   | no (0)      | yes (1)  | `101`                 | `["subject", "object"]`              |
| no (0)    | no (0)      | no (0)   | `000`                 | `[]`                                 |
| yes (1)   | yes (1)     | yes (1)  | `111`                 | `["subject", "predicate", "object"]` |

What the code is doing here is generating the numbers `001` to `110` [^1] and creating and index for each number with the respective key path. I cannot think of any simpler way to do that.

[^1]: No need to create indexes for `000` and `111` as for those queries the object store itself can be used.

Reply:
I got the logic now :D My point is that it was a bit hard to understand because I didn't know what this algorithm was trying to do, i.e. generate all possible subsets of `QUAD_KEYS`. Knowing that - through moving it to a separate function or even a comment, I would have gotten it faster.